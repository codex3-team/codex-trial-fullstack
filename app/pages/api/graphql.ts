import { ApolloServer, gql, makeExecutableSchema } from 'apollo-server-micro'
import { getKnex } from '../../knex'

const typeDefs = gql`
  type Query {
    total: Int
    cars(offset: Int, limit: Int): [Car!]!
  }
  type Mutation {
    insertCar(make: String!, model: String!, year: String!): Car!
  }
  type Car {
    id: ID!
    make: String
    model: String
    year: String
  }
`
const knex = getKnex()
const HOUR = 1 * 1000 * 60 * 60

// TS isn't fully supported for memory-cache
const cache = require('memory-cache')
const resolvers = {
  Query: {
    total(_parent, _args, _ctx, _info) {
      const cached = cache.get('total')

      if (cached) {
        return Promise.resolve(cached)
      } else {
        return knex
        .table(process.env.TABLE_NAME)
        .count()
        .then((row: any) => {
          const total = row[0].count
          cache.put('total', total, HOUR)

          return total
        })
      }
    },
    cars(_parent, args, _ctx, _info) {
      return knex
        .table(process.env.TABLE_NAME)
        .select()
        .orderBy('year', 'desc')
        .limit(args.limit || process.env.NEXT_PUBLIC_SIZE)
        .offset(args.offset || 0)
    }
  },
  Mutation: {
    insertCar(_parent, args, _ctx, _info) {
      return knex
        .insert(args)
        .returning('id')
        .into(process.env.TABLE_NAME)
        .then((res: any) => {
          cache.del('total')

          return { id: res[0] }
        })
    }
  }
}
const apolloServer = new ApolloServer({
  typeDefs, resolvers, uploads: false
})

export const config = {
  api: {
    bodyParser: false
  }
}

export default apolloServer.createHandler({ path: '/api/graphql' })
export const schema = makeExecutableSchema({ typeDefs, resolvers })
