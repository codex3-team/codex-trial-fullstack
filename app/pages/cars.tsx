import { useEffect, useState } from 'react'
import { useRouter } from 'next/router'
import Car from '../components/Car'
import Nav from '../components/nav'
import Footer from '../components/Footer'
import Menu from '../components/Menu'

function Cars() {
    const [cars, setCars] = useState([])
    const [total, setTotal] = useState(0)

    const router = useRouter()
    const p = Number(router.query.p || 1)

    useEffect(() => {
        async function loadCars() {
            const resp = await fetch(`/api/cars?p=${p}`)
            const results = await resp.json()

            if (results.errors && results.errors.length) {
                console.warn(JSON.stringify(results.errors))
                setCars([])
                setTotal(0)
            } else if (results.data && results.data.cars) {
                setCars(results.data.cars)
                setTotal(results.data.total)
            }
        }
        loadCars()
    }, [p])

    return (
        <div className="flex flex-col min-h-screen">
            <Menu></Menu>

            <main className="flex flex-col items-center flex-1 px-20">
                <h1 className="text-6xl font-bold py-10">Cars</h1>
                <div className="bg-white overflow-hidden">
                    <div className="grid grid-cols-3 gap-4">
                    {!total && <small>loading...</small>}
                    {!!total && cars.map((car) => (
                        <Car key={car.id} id={car.id} model={car.model} make={car.make} year={car.year}></Car>
                    ))}
                    </div>
                </div>
            </main>

            <Nav total={total}></Nav>
            <Footer></Footer>
        </div>
    )
}

export default Cars
