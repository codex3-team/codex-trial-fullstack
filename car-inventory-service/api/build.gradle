plugins {
    id 'java-library'
}

version = '0.0.1-SNAPSHOT'
archivesBaseName = 'codex-api'

dependencies {
    implementation "org.springframework.boot:spring-boot-starter-web"
    implementation "org.springdoc:springdoc-openapi-ui:1.5.9"
}