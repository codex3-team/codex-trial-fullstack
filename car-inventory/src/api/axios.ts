import axios from 'axios'

export const codexContextPath = '/codex/api'

const axiosInstance = axios.create()

axiosInstance.interceptors.request.use(
    (request) => requestHandler(request),
    (error) => errorHandler(error)
)

const errorHandler = (error: any) => {
    return Promise.reject(error)
}

const requestHandler = async (request: any) => {
    request.baseURL = 'http://localhost:8008'
    return request
}

export default axiosInstance