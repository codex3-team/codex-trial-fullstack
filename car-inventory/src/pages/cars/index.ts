import CarsPage from './CarsPage'
export { CarsPage }