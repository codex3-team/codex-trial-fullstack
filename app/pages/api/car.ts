import { graphql } from 'graphql'
import { schema } from './graphql'

const insertCar = `
    mutation($make: String!, $model: String!, $year: String!) {
        insertCar(make: $make, model: $model, year: $year) {
            id
        }
    }
`

export default async function handler(req, res) {
    if (req.method == 'POST') {
        const car = await graphql({ schema, source: insertCar, variableValues: JSON.parse(req.body) })
        return res.status(200).json(car)
    }
}
