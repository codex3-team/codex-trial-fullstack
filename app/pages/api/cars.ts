import { graphql } from 'graphql'
import { schema } from './graphql'

export const queryCars = `
    query($offset: Int, $limit: Int) {
        total
        cars(offset: $offset, limit: $limit) {
            id
            model
            make
            year
        }
    }
`

export default async function handler(req, res) {
    const p = Number(req.query.p || 1)
    const size = Number(process.env.NEXT_PUBLIC_SIZE)
    const vars = {
        limit: size,
        offset: (p - 1) * size
    }

    const cars = await graphql({ schema, source: queryCars, variableValues: vars })
    return res.status(200).json(cars)
}
