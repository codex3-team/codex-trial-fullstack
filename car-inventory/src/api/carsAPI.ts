import { AxiosResponse } from 'axios'
import { Car } from '../types/car.interface'
import { CarPayload } from '../types/carPayload.interface'
import { CarResponse } from '../types/carResponse.interface'
import instance, { codexContextPath } from './axios'

const responseBody = (response: AxiosResponse) => response.data

export const getCars = (carPayload: CarPayload): Promise<CarResponse> => {
    const { page, size, sort } = carPayload
    return instance.get(`${codexContextPath}/v1/cars?page=${page}&size=${size}&sort=${sort}`).then(responseBody)
}

export const postCar = (car: Car): Promise<AxiosResponse> =>
    instance.post(`${codexContextPath}/v1/cars`, car)