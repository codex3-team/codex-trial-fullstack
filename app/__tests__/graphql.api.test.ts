import 'regenerator-runtime/runtime'
import { createTestClient } from 'apollo-server-testing'
import { ApolloServer } from 'apollo-server-micro'
import { schema } from '../pages/api/graphql'
import { queryCars } from '../pages/api/cars'

const mocks = {
    Query: () =>({
        total: () => (1),
        cars: () => ([{ id: '1234-5678', model: 'Model', make: 'Ford', year: '1950' }])
    })
}
const server = new ApolloServer({
    schema: schema,
    mocks: mocks
})

describe('Graphql API', () => {
    it('fetches cars', async () => {
        const { query } = createTestClient(server)

        const res = await query({
            query: queryCars,
            variables: {
                limit: 1,
                offset: 0
            }
        })
        expect(res).toMatchSnapshot()
    })
})
