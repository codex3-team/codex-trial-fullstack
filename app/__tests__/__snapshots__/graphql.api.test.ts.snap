// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graphql API fetches cars 1`] = `
Object {
  "data": Object {
    "cars": Array [
      Object {
        "id": "1234-5678",
        "make": "Ford",
        "model": "Model",
        "year": "1950",
      },
    ],
    "total": 1,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
